### ISSUE TITLE
Improvement for exceptions "Concurrent modification"

### ISSUE THREAD

**Issue by QiXi** [dart-lang/language#1720]
the following describes how the direction of collection iteration affects the writing of stable code

```
final Map<SocketId, GameClient> _clientById = {};
...
void disconnect() {
 for (var client in _clientById.values) {
  client.disconnect(false); // internal implementation calls _clientById.remove(element)  <-- Unhandled exception: Concurrent modification during iteration: _LinkedHashMap len:0.
 }
 ...
}
```

There are several ways to prevent the error "Concurrent modification".
I know the only way that does not use auxiliary collections. You need to go through the list in the opposite direction.

```
void disconnect() {
 for (var i = _clientById.length - 1; i >= 0; i--) {
 var client = _clientById.values.elementAt(i);
  client.disconnect(false); //<-- now the code is safe and has no errors
 }
}
```

we have to write a lot of code for a simple iteration in the opposite direction

I would like to be able to traverse in the same simple form as direct traversal and without generating additional collections
for example it could be like this
```
 forReverse (var client in _clientById.values) {
  ...
 }
```
or
```
 for (var client in _clientById.values.reverse) {
  ...
 }
```
or another 

this improvement is better viewed not from the point of view of convenience, but from the point of view of writing safe code

_special thanks for the "null safety" in the Dart_

**Comment by mateusfccp** [comment]
I didn't test, but doesn't this solves your problem?

```dart
for (var client in _clientById.values.toList().reversed) {
  // ...
}
```

**Comment by QiXi** [comment]
my years of Java experience will not allow me to create a new List in a place where I can do without spawning a new collection. in my opinion it is very wasteful. At the moment I do not have experience in writing production code in Dart and do not know all the subtleties that can save performance, so I am focusing on my experience from Java.

**Comment by mateusfccp** [comment]
Yes, using `.toList()` may give a little overhead, but it will negligible.

Anyway, I don't think we could bring `.reversed` up to `Iterable`, considering its structure.

Also, I think this discussion belongs to `dart-lang/sdk` instead of `dart-lang/language`.

**Comment by QiXi** [comment]
in fact, it's not difficult for me to write a few extra lines to solve the problem.
My biggest concern is that the default code is prone to "Concurrent modification" errors. In a large library of complex code, it is not always obvious that iterating over a list can cause an item to be removed from the list. Perhaps the area for solving these problems can be in a completely different cotext. I just took the opportunity to express my concern. Sorry for the English messages from the translator.

**Comment by QiXi** [comment]
in fact, it's not difficult for me to write a few extra lines to solve the problem.
My biggest concern is that the default code is prone to "Concurrent modification" errors. In a large library of complex code, it is not always obvious that iterating over a list can cause an item to be removed from the list. Perhaps the area for solving these problems can be in a completely different cotext. I just took the opportunity to express my concern. Sorry for the English messages from the translator.
